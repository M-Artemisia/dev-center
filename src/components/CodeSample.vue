<template>
  <div class="syn-code-block">
    <div class="syn-code-block-header syn-flex justify-space-between">
      <span class="syn-overline syn-mb-0">{{title}}</span>
      <div class="syn-ma-8 syn-flex align-center">
        <select class="syn-code-language-selector syn-mr-8" v-if="languages.length > 1" v-model="selectedLanguage">
          <option v-for="language in languages" :key="language" :value="language">
            {{ language }}
          </option>
        </select>
        <button class="syn-btn no-padding" @click="copySample" title="Copy">
          <svg width="14" height="14" viewBox="0 0 24 24" class="syn-ma-4">
            <path
              fill="currentColor"
              d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm-1 4H8c-1.1 0-1.99.9-1.99 2L6 21c0 1.1.89 2 1.99 2H19c1.1 0 2-.9 2-2V11l-6-6zM8 21V7h6v5h5v9H8z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div class="syn-code-block-body" ref="codeContent">
      <slot>
        <pre><code>// Sample Code Block
  console.log('Hello world!')</code></pre>
      </slot>
    </div>
  </div>
</template>

<script>
export default {
  name: 'CodeSample',
  props: {
    title: {
      type: String,
      default: ''
    }
  },
  data() {
    return {
      languages: [],
      selectedLanguage: ''
    }
  },
  methods: {
    copySample() {
      const copyText = this.$refs.codeContent.querySelector(`figure.${this.selectedLanguage}`).innerText
      const el = document.createElement('textarea')
      el.value = copyText
      document.body.appendChild(el)
      el.select()
      document.execCommand('copy')
      document.body.removeChild(el)
      this.$root.$emit('show-toast', 'Copied to clipboard!')
    },
    processLanguages() {
      const samples = this.$refs.codeContent.querySelectorAll('figure')
      const langs = []
      samples.forEach((sample) => {
        let sampleLang = ''
        if (sample.parentElement.getAttribute('code-sample-language')) {
          // First check if code-sample-language was specified and use that if it was
          sampleLang = sample.parentElement.getAttribute('code-sample-language')
        } else {
          // Otherwise default to the language specified on the highlight block
          const sampleCode = sample.querySelector('code')
          sampleLang = sampleCode.getAttribute('data-lang')
        }
        langs.push(sampleLang)
        sample.classList.add(sampleLang)
      })
      return langs
    },
    showSelectedLanguage() {
      const samples = this.$refs.codeContent.querySelectorAll('figure')
      samples.forEach((sample) => {
        if (sample.classList.contains(this.selectedLanguage)) {
          sample.hidden = false
        } else {
          sample.hidden = true
        }
      })
    }
  },
  mounted() {
    // Wait to check for languages till mounted
    this.languages = this.processLanguages()
    if (this.languages.length) this.selectedLanguage = this.languages[0]
  },
  watch: {
    selectedLanguage(newVal, oldVal) {
      if (newVal !== oldVal) this.showSelectedLanguage()
    }
  }
}
</script>
