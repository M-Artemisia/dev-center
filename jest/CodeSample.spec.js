import { createLocalVue, mount } from '@vue/test-utils'
import Vuex from 'vuex'
import CodeSample from '@/components/CodeSample'
import nock from 'nock'

describe('CodeSample', () => {
  const state = {}
  const getters = {
    preferredLanguage: () => ''
  }
  const actions = {
    setPreferredLanguage: () => {}
  }
  const localVue = createLocalVue()
  localVue.use(Vuex)
  const store = new Vuex.Store({ state, getters, actions })


  test('should render default slot content if no custom content provided', () => {
    const wrapper = mount(CodeSample, {
      localVue,
      store,
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render provided content in slot', () => {
    const wrapper = mount(CodeSample, {
      localVue,
      store,
      slots: {
        default: `<figure class="highlight"><pre><code data-lang="bash" class="language-bash">POST https://api.algorithmia.com/v1/algo/:owner/:algoname</code></pre></figure>`
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should show language dropdown and hide additional figures if multiple provided', () => {
    const wrapper = mount(CodeSample, {
      localVue,
      store,
      slots: {
        default: `<figure class="highlight python"><pre><code data-lang="python" class="language-python"><span class="kn">import</span> <span class="nn">Algorithmia</span></code></pre></figure>

        <figure class="highlight"><pre><code data-lang="javascript" class="language-javascript"><span class="kd">var</span> <span class="nx">input</span> <span class="o">=</span> <span class="s2">"YOUR_USERNAME"</span><span class="p">;</span></code></pre></figure>`
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should change code sample to selected language if changed', () => {
    const wrapper = mount(CodeSample, {
      localVue,
      store,
      slots: {
        default: `<figure class="highlight python"><pre><code data-lang="python" class="language-python"><span class="kn">import</span> <span class="nn">Algorithmia</span></code></pre></figure>

        <figure class="highlight"><pre><code data-lang="javascript" class="language-javascript"><span class="kd">var</span> <span class="nx">input</span> <span class="o">=</span> <span class="s2">"YOUR_USERNAME"</span><span class="p">;</span></code></pre></figure>`
      }
    })
    wrapper.find('select').setValue('javascript')
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should use code-sample-language instead of data-lang if provided', () => {
    const wrapper = mount(CodeSample, {
      localVue,
      store,
      slots: {
        default: `<figure class="highlight python"><pre><code data-lang="python" class="language-python"><span class="kn">import</span> <span class="nn">Algorithmia</span></code></pre></figure>

        <div code-sample-language="JavaScript"><figure class="highlight"><pre><code data-lang="javascript" class="language-javascript"><span class="kd">var</span> <span class="nx">input</span> <span class="o">=</span> <span class="s2">"YOUR_USERNAME"</span><span class="p">;</span></code></pre></figure></div>`
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
