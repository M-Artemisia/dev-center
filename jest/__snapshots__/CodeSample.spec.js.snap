// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeSample should change code sample to selected language if changed 1`] = `
<div class="syn-code-block">
  <div class="syn-code-block-header syn-flex justify-space-between align-center"><span class="syn-overline syn-text-light-primary syn-mb-0 syn-ml-16 syn-mr-8"></span>
    <div class="syn-ma-8 syn-flex align-center"><select class="syn-code-language-selector syn-mr-8">
        <option value="python">
          python
        </option>
        <option value="javascript">
          javascript
        </option>
      </select> <button title="Copy" class="syn-btn no-padding"><i class="material-icons syn-ma-4">content_copy</i></button></div>
  </div>
  <div class="syn-code-block-body">
    <figure class="highlight python" hidden=""><pre><code data-lang="python" class="language-python"><span class="kn">import</span> <span class="nn">Algorithmia</span></code></pre>
    </figure>
    <figure class="highlight javascript"><pre><code data-lang="javascript" class="language-javascript"><span class="kd">var</span> <span class="nx">input</span> <span class="o">=</span> <span class="s2">"YOUR_USERNAME"</span><span class="p">;</span></code></pre>
    </figure>
  </div>
</div>
`;

exports[`CodeSample should render default slot content if no custom content provided 1`] = `
<div class="syn-code-block">
  <div class="syn-code-block-header syn-flex justify-space-between align-center"><span class="syn-overline syn-text-light-primary syn-mb-0 syn-ml-16 syn-mr-8"></span>
    <div class="syn-ma-8 syn-flex align-center">
      <!----> <button title="Copy" class="syn-btn no-padding"><i class="material-icons syn-ma-4">content_copy</i></button></div>
  </div>
  <div class="syn-code-block-body"><pre><code>// Sample Code Block
console.log('Hello world!')</code></pre>
  </div>
</div>
`;

exports[`CodeSample should render provided content in slot 1`] = `
<div class="syn-code-block">
  <div class="syn-code-block-header syn-flex justify-space-between align-center"><span class="syn-overline syn-text-light-primary syn-mb-0 syn-ml-16 syn-mr-8"></span>
    <div class="syn-ma-8 syn-flex align-center">
      <!----> <button title="Copy" class="syn-btn no-padding"><i class="material-icons syn-ma-4">content_copy</i></button></div>
  </div>
  <div class="syn-code-block-body">
    <figure class="highlight bash"><pre><code data-lang="bash" class="language-bash">POST https://api.algorithmia.com/v1/algo/:owner/:algoname</code></pre>
    </figure>
  </div>
</div>
`;

exports[`CodeSample should show language dropdown and hide additional figures if multiple provided 1`] = `
<div class="syn-code-block">
  <div class="syn-code-block-header syn-flex justify-space-between align-center"><span class="syn-overline syn-text-light-primary syn-mb-0 syn-ml-16 syn-mr-8"></span>
    <div class="syn-ma-8 syn-flex align-center"><select class="syn-code-language-selector syn-mr-8">
        <option value="python">
          python
        </option>
        <option value="javascript">
          javascript
        </option>
      </select> <button title="Copy" class="syn-btn no-padding"><i class="material-icons syn-ma-4">content_copy</i></button></div>
  </div>
  <div class="syn-code-block-body">
    <figure class="highlight python"><pre><code data-lang="python" class="language-python"><span class="kn">import</span> <span class="nn">Algorithmia</span></code></pre>
    </figure>
    <figure class="highlight javascript" hidden=""><pre><code data-lang="javascript" class="language-javascript"><span class="kd">var</span> <span class="nx">input</span> <span class="o">=</span> <span class="s2">"YOUR_USERNAME"</span><span class="p">;</span></code></pre>
    </figure>
  </div>
</div>
`;

exports[`CodeSample should use code-sample-language instead of data-lang if provided 1`] = `
<div class="syn-code-block">
  <div class="syn-code-block-header syn-flex justify-space-between align-center"><span class="syn-overline syn-text-light-primary syn-mb-0 syn-ml-16 syn-mr-8"></span>
    <div class="syn-ma-8 syn-flex align-center"><select class="syn-code-language-selector syn-mr-8">
        <option value="python">
          python
        </option>
        <option value="JavaScript">
          JavaScript
        </option>
      </select> <button title="Copy" class="syn-btn no-padding"><i class="material-icons syn-ma-4">content_copy</i></button></div>
  </div>
  <div class="syn-code-block-body">
    <figure class="highlight python"><pre><code data-lang="python" class="language-python"><span class="kn">import</span> <span class="nn">Algorithmia</span></code></pre>
    </figure>
    <div code-sample-language="JavaScript">
      <figure class="highlight JavaScript" hidden=""><pre><code data-lang="javascript" class="language-javascript"><span class="kd">var</span> <span class="nx">input</span> <span class="o">=</span> <span class="s2">"YOUR_USERNAME"</span><span class="p">;</span></code></pre>
      </figure>
    </div>
  </div>
</div>
`;
